flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java

```diff
@@ -473,25 +521,40 @@ public T deserialize(T reuse, DataInputView source) throws IOException {
             }
         }
 
-        if ((flags & NO_SUBCLASS) != 0) {
+        if (isRecord()) {
             try {
+                JavaRecordBuilderFactory<T>.JavaRecordBuilder builder = recordHelper.newBuilder();
                 for (int i = 0; i < numFields; i++) {
                     boolean isNull = source.readBoolean();
 
                     if (fields[i] != null) {
                         if (isNull) {
-                            fields[i].set(reuse, null);
+                            builder.setField(i, null);
                         } else {
-                            Object field;
+                            builder.setField(i, deserializeField(null, i, source));
```

Comment:
let me just revert this change, reuse may or may not be null. If you pass a non-null reuse it's not null, but otherwise createInstance returns null for records

Reply:
> If you pass a non-null reuse it's not null, but otherwise createInstance returns null for records

true, I overlooked that `createInstance` actually returns `null` in certain situations.